src/box/lua/config/applier/box_cfg.lua

```diff
@@ -475,7 +444,141 @@ local function apply(config)
+    fiber.self():name(name)
+
+    local version = mkversion.get()
+    if version < mkversion(3, 0, 0) then
+        fiber.sleep(TIMEOUT_INFINITY)
+    end
+
+    box.ctl.wait_rw()
+
+    -- Somebody else already applied all names.
+    if table.equals(configdata._missing_names, {}) then
+        return
+    end
+
+    box.begin()
+    for name, uuid in pairs(configdata._missing_names) do
+        if name == configdata:names().replicaset_name then
+            box.space._schema:insert{'replicaset_name', name}
+        else
+            local tuple = box.space._cluster.index.uuid:select(uuid)
+            box.space._cluster:update(tuple[1][1], {{'=', 3, name}})
+        end
+    end
+    box.commit()
+end
+
+local function names_schema_on_replace(old, new)
+    if old == nil and new[1] == 'replicaset_name' then
+        box.on_commit(function()
+            local watch_key = 'box_apply_names'
+            local names = box.watch_once(watch_key) or {}
+            table.insert(names, new[2])
+            box.broadcast(watch_key, names)
+        end)
+    end
+end
+
+local function names_cluster_on_replace(old, new)
+    if old[3] == nil and new[3] ~= nil then
+        box.on_commit(function()
+            local watch_key = 'box_apply_names'
+            local names = box.watch_once(watch_key) or {}
+            table.insert(names, new[3])
+            box.broadcast(watch_key, names)
+        end)
+    end
+end
+
+local function names_box_cfg(cfg)
+    local ok, err = pcall(function()
+        box.cfg(cfg)
+    end)
+
+    -- We don't consider failed box.cfg a critical error,
+    -- delete a warning anyway. box.info shows correct info.
+    if not ok then
+        log.warn('Failed to apply name in box.cfg: %s', err)
+    end
+end
+
+local function post_apply(config)
+    local configdata = config._configdata
+    if table.equals(configdata._missing_names, {}) then
+        return
+    end
+
+    local names_apply_fiber = fiber.new(names_apply_f, configdata)
+
+    -- Set triggers to notify apply_name watcher
+    box.space._cluster:on_replace(names_cluster_on_replace)
+    box.space._schema:on_replace(names_schema_on_replace)
+
+    local upgrade_on_replace = nil
+    local version = mkversion.get()
+    if version < mkversion(3, 0, 0) then
+        upgrade_on_replace = box.space._schema:on_replace(function(old, new)
+            if old == nil or new == nil then
+                return
+            end
+
+            local version_3 = mkversion(3, 0, 0)
+            local old_version = mkversion.from_tuple(old)
+            local new_version = mkversion.from_tuple(new)
+            if old_version < version_3 and new_version >= version_3 then
+                box.on_commit(function()
+                    pcall(names_apply_fiber.wakeup, names_apply_fiber)
+                end)
+            end
+        end)
+    end
+
+    local watch
+    local config_names = configdata:names()
+    watch = box.watch('box_apply_names', function(_, names)
```

Comment:
Consider doing this with `trigger` module, proposed by @Totktonada

Reply:
We cannot do this using trigger in or_replace/on_commit, as it's synchronously executed. We need asynchronous event loop for that